Async version of `list` command?

I'm trying to use `rust-s3` in a method returning a Future to asynchronously get data from an S3 bucket and am currently using version `0.16.0-alpha2`. Using `get_object_async` is already working fine for me, however, I would now like to list the contents of the Bucket to download further objects in the same async function. However, if I call `mybucket.list`, it seems that the list method is not currently compatible with asynchronous operation:

```
executor failed to spawn task: tokio::spawn failed (is a tokio runtime running this future?)
```

Would it be possible to add a `list_async` function similar to the `get_object_async` method?

Many thanks for providing this very useful library!

OK, I've tried now for a while but am not sure if there is a way to get this to work without altering `rust-s3`. Here is my minimal example:

```rust
fn get_bucket_list<'a>(
    bucket: &'a s3::bucket::Bucket,
    path: &'a str,
) -> Box<dyn Future<Item = Vec<String>, Error = Error> + Send> {
    let fut = bucket
        .list_all_async(path, Some("/"))
        .map(|res| {
            res.into_iter()
                .flat_map(|a| a.contents.into_iter().map(|o| o.key.clone()))
                .collect()
        })
        .map_err(|_e| format_err!("Error listing bucket"));

    Box::new(fut)
}
```
I need to return a `Future` that is also `Send` because I want to return it to an actix-web route handler. The code above refuses to compile with the following longish error message:

```
error[E0277]: `<impl futures::future::Future as futures::future::Future>::Error` cannot be sent between threads safely
   --> src/source.rs:139:5
    |
139 |     Box::new(fut)
    |     ^^^^^^^^^^^^^ `<impl futures::future::Future as futures::future::Future>::Error` cannot be sent between threads safely
    |
    = help: within `futures::future::or_else::OrElse<futures::future::map::Map<impl futures::future::Future, [closure@src/source.rs:132:14: 136:10]>, std::result::Result<std::vec::Vec<std::string::String>, failure::error::Error>, [closure@src/source.rs:137:18: 137:37]>`, the trait `std::marker::Send` is not implemented for `<impl futures::future::Future as futures::future::Future>::Error`
    = note: required because it appears within the type `std::result::Result<std::vec::Vec<s3::serde_types::ListBucketResult>, <impl futures::future::Future as futures::future::Future>::Error>`
    = note: required because it appears within the type `std::option::Option<std::result::Result<std::vec::Vec<s3::serde_types::ListBucketResult>, <impl futures::future::Future as futures::future::Future>::Error>>`
    = note: required because it appears within the type `futures::future::result_::FutureResult<std::vec::Vec<s3::serde_types::ListBucketResult>, <impl futures::future::Future as futures::future::Future>::Error>`
    = note: required because it appears within the type `futures::future::chain::Chain<futures::future::loop_fn::LoopFn<futures::future::and_then::AndThen<impl futures::future::Future, std::result::Result<futures::future::loop_fn::Loop<(std::option::Option<&str>, std::vec::Vec<s3::serde_types::ListBucketResult>), (std::option::Option<std::string::String>, std::vec::Vec<s3::serde_types::ListBucketResult>)>, <impl futures::future::Future as futures::future::Future>::Error>, [closure@DefId(166:84 ~ s3[9bed]::bucket[0]::{{impl}}[0]::list_all_async[0]::{{closure}}[0]::{{closure}}[0]) 0:std::vec::Vec<s3::serde_types::ListBucketResult>]>, [closure@DefId(166:83 ~ s3[9bed]::bucket[0]::{{impl}}[0]::list_all_async[0]::{{closure}}[0]) 0:&s3::bucket::Bucket, 1:&str, 2:std::option::Option<&str>]>, futures::future::result_::FutureResult<std::vec::Vec<s3::serde_types::ListBucketResult>, <impl futures::future::Future as futures::future::Future>::Error>, [closure@DefId(166:85 ~ s3[9bed]::bucket[0]::{{impl}}[0]::list_all_async[0]::{{closure}}[1])]>`
    = note: required because it appears within the type `futures::future::and_then::AndThen<futures::future::loop_fn::LoopFn<futures::future::and_then::AndThen<impl futures::future::Future, std::result::Result<futures::future::loop_fn::Loop<(std::option::Option<&str>, std::vec::Vec<s3::serde_types::ListBucketResult>), (std::option::Option<std::string::String>, std::vec::Vec<s3::serde_types::ListBucketResult>)>, <impl futures::future::Future as futures::future::Future>::Error>, [closure@DefId(166:84 ~ s3[9bed]::bucket[0]::{{impl}}[0]::list_all_async[0]::{{closure}}[0]::{{closure}}[0]) 0:std::vec::Vec<s3::serde_types::ListBucketResult>]>, [closure@DefId(166:83 ~ s3[9bed]::bucket[0]::{{impl}}[0]::list_all_async[0]::{{closure}}[0]) 0:&s3::bucket::Bucket, 1:&str, 2:std::option::Option<&str>]>, std::result::Result<std::vec::Vec<s3::serde_types::ListBucketResult>, <impl futures::future::Future as futures::future::Future>::Error>, [closure@DefId(166:85 ~ s3[9bed]::bucket[0]::{{impl}}[0]::list_all_async[0]::{{closure}}[1])]>`
    = note: required because it appears within the type `impl futures::future::Future`
    = note: required because it appears within the type `futures::future::map::Map<impl futures::future::Future, [closure@src/source.rs:132:14: 136:10]>`
    = note: required because it appears within the type `futures::future::chain::Chain<futures::future::map::Map<impl futures::future::Future, [closure@src/source.rs:132:14: 136:10]>, futures::future::result_::FutureResult<std::vec::Vec<std::string::String>, failure::error::Error>, [closure@src/source.rs:137:18: 137:37]>`
    = note: required because it appears within the type `futures::future::or_else::OrElse<futures::future::map::Map<impl futures::future::Future, [closure@src/source.rs:132:14: 136:10]>, std::result::Result<std::vec::Vec<std::string::String>, failure::error::Error>, [closure@src/source.rs:137:18: 137:37]>`
    = note: required for the cast to the object type `dyn futures::future::Future<Error = failure::error::Error, Item = std::vec::Vec<std::string::String>> + std::marker::Send`
```

if it's any use, `Error` is `failure::Error`. It looks like some `Error` associated type is not `Send` making the rest of the `Future` not `Send`.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "rust-s3"
-version = "0.18.0-beta1"
+version = "0.18.0-beta2"
 authors = ["Drazen Urch", "Nick Stevens"]
 description = "Tiny Rust library for working with Amazon S3."
 repository = "https://github.com/durch/rust-s3"

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "rust-s3"
-version = "0.18.0-beta"
+version = "0.18.0-beta1"
 authors = ["Drazen Urch", "Nick Stevens"]
 description = "Tiny Rust library for working with Amazon S3."
 repository = "https://github.com/durch/rust-s3"

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "rust-s3"
-version = "0.18.0-beta2"
+version = "0.18.0-beta3"
 authors = ["Drazen Urch", "Nick Stevens"]
 description = "Tiny Rust library for working with Amazon S3."
 repository = "https://github.com/durch/rust-s3"

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "rust-s3"
-version = "0.18.0-beta3"
+version = "0.18.0-beta4"
 authors = ["Drazen Urch", "Nick Stevens"]
 description = "Tiny Rust library for working with Amazon S3."
 repository = "https://github.com/durch/rust-s3"
@@ -28,7 +28,6 @@ url = "^1.5.1"
 rust-ini = "^0.9"
 dirs = "^1.0"
 futures = "^0.1.28"
-snafu = {version = "^0.5", features = ["futures-01"]}
 
 [features]
 no-verify-ssl = []

--- a/src/bin/simple_crud.rs
+++ b/src/bin/simple_crud.rs
@@ -1,35 +1,20 @@
 extern crate s3;
-extern crate snafu;
-
-use snafu::{ResultExt, Snafu};
 
 use std::str;
 
 use s3::bucket::Bucket;
 use s3::credentials::Credentials;
+use s3::error::S3Error;
 
 const BUCKET: &str = "drazen-test-bucket-2";
 const MESSAGE: &str = "I want to go to S3";
 const REGION: &str = "us-east-1";
 
-#[derive(Debug, Snafu)]
-pub enum Error {
-    InvalidRegion { source: s3::region::Error },
-    BucketCreate { source: s3::bucket::Error },
-    BucketPut { source: s3::bucket::Error },
-    BucketGet { source: s3::bucket::Error },
-    BucketPutTag { source: s3::bucket::Error },
-    BucketGetTag { source: s3::bucket::Error },
-    BucketLocation { source: s3::bucket::Error },
-}
-
-type S3Result<T, E = Error> = std::result::Result<T, E>;
-
-pub fn main() -> S3Result<()> {
-    let region = REGION.parse().context(InvalidRegion)?;
+pub fn main() -> Result<(), S3Error> {
+    let region = REGION.parse()?;
 //     Create Bucket in REGION for BUCKET
     let credentials = Credentials::default();
-    let bucket = Bucket::new(BUCKET, region, credentials).context(BucketCreate)?;
+    let bucket = Bucket::new(BUCKET, region, credentials)?;
 
     // List out contents of directory
 //    let results = bucket.list("", None).unwrap();
@@ -48,22 +33,22 @@ pub fn main() -> S3Result<()> {
 
     // Put a "test_file" with the contents of MESSAGE at the root of the
     // bucket.
-    let (_, code) = bucket.put_object("test_file", MESSAGE.as_bytes(), "text/plain").context(BucketPut)?;
+    let (_, code) = bucket.put_object("test_file", MESSAGE.as_bytes(), "text/plain")?;
     assert_eq!(200, code);
 
     // Get the "test_file" contents and make sure that the returned message
     // matches what we sent.
-    let (data, code) = bucket.get_object("test_file").context(BucketGet)?;
+    let (data, code) = bucket.get_object("test_file")?;
     let string = str::from_utf8(&data).unwrap();
     assert_eq!(200, code);
     assert_eq!(MESSAGE, string);
 
 //  Get bucket location
-    println!("{:?}", bucket.location().context(BucketLocation)?);
+    println!("{:?}", bucket.location()?);
 
-    bucket.put_object_tagging("test_file", &[("test", "tag")]).context(BucketPutTag)?;
+    bucket.put_object_tagging("test_file", &[("test", "tag")])?;
     println!("Tags set");
-    let (tags, _status) = bucket.get_object_tagging("test_file").context(BucketGetTag)?;
+    let (tags, _status) = bucket.get_object_tagging("test_file")?;
     println!("{:?}", tags);
 
     Ok(())

--- a/src/bucket.rs
+++ b/src/bucket.rs
@@ -7,28 +7,13 @@ use std::io::Write;
 use credentials::Credentials;
 use command::Command;
 use region::Region;
-use request::{Request, Headers, Query, S3Error};
+use request::{Request, Headers, Query};
 use serde_types::{ListBucketResult, BucketLocationResult};
 use serde_types::Tagging;
 use futures::Future;
-use futures::future::{loop_fn, Loop, ok, result, LoopFn};
-use snafu::{ResultExt, Snafu};
+use futures::future::{loop_fn, Loop};
+use error::{S3Error, S3Result};
 
-#[derive(Debug, Snafu)]
-pub enum Error {
-    XmlError {
-        source: serde_xml::Error
-    },
-    RequestError {
-        source: ::request::Error
-    },
-    ParseError {
-        source: ::region::Error
-    },
-    ResponseError
-}
-
-type S3Result<T, E = Error> = std::result::Result<T, E>;
 
 /// # Example
 /// ```
@@ -101,7 +86,7 @@ impl Bucket {
     pub fn get_object(&self, path: &str) -> S3Result<(Vec<u8>, u16)> {
         let command = Command::GetObject;
         let request = Request::new(self, path, command);
-        Ok(request.response_data().context(RequestError)?)
+        Ok(request.response_data()?)
     }
 
     /// Gets file from an S3 path.
@@ -153,7 +138,7 @@ impl Bucket {
     pub fn get_object_stream<T: Write>(&self, path: &str, writer: &mut T) -> S3Result<u16> {
         let command = Command::GetObject;
         let request = Request::new(self, path, command);
-        Ok(request.response_data_to_writer(writer).context(RequestError)?)
+        Ok(request.response_data_to_writer(writer)?)
     }
 
     /// Stream file from S3 path to a local file, generic over T: Write.
@@ -206,18 +191,18 @@ impl Bucket {
     /// ```
     pub fn location(&self) -> S3Result<(Region, u16)> {
         let request = Request::new(self, "?location", Command::GetBucketLocation);
-        let result = request.response_data().context(RequestError)?;
+        let result = request.response_data()?;
         let region_string = String::from_utf8_lossy(&result.0);
         let region = match serde_xml::deserialize(region_string.as_bytes()) {
             Ok(r) => {
                 let location_result: BucketLocationResult = r;
-                location_result.region.parse().context(ParseError)?
+                location_result.region.parse()?
             }
             Err(e) => {
                 if e.to_string() == "missing field `$value`" {
-                    "us-east-1".parse().context(ParseError)?
+                    "us-east-1".parse()?
                 } else {
-                    panic!("lala")
+                    panic!("How did we get here?")
                 }
             }
         };
@@ -243,7 +228,7 @@ impl Bucket {
     pub fn delete_object(&self, path: &str) -> S3Result<(Vec<u8>, u16)> {
         let command = Command::DeleteObject;
         let request = Request::new(self, path, command);
-        request.response_data().context(RequestError)
+        request.response_data()
     }
 
     /// Put into an S3 bucket.
@@ -273,7 +258,7 @@ impl Bucket {
             content_type,
         };
         let request = Request::new(self, path, command);
-        Ok(request.response_data().context(RequestError)?)
+        Ok(request.response_data()?)
     }
 
     fn _tags_xml(&self, tags: &[(&str, &str)]) -> String {
@@ -317,7 +302,7 @@ impl Bucket {
             tags: &content.to_string()
         };
         let request = Request::new(self, path, command);
-        Ok(request.response_data().context(RequestError)?)
+        Ok(request.response_data()?)
     }
 
 
@@ -344,7 +329,7 @@ impl Bucket {
     pub fn delete_object_tagging(&self, path: &str) -> S3Result<(Vec<u8>, u16)> {
         let command = Command::DeleteObjectTagging;
         let request = Request::new(self, path, command);
-        Ok(request.response_data().context(RequestError)?)
+        Ok(request.response_data()?)
     }
 
     /// Retrieve an S3 object list of tags.
@@ -374,12 +359,12 @@ impl Bucket {
     pub fn get_object_tagging(&self, path: &str) -> S3Result<(Option<Tagging>, u16)> {
         let command = Command::GetObjectTagging {};
         let request = Request::new(self, path, command);
-        let result = request.response_data().context(RequestError)?;
+        let result = request.response_data()?;
 
         let tagging = if result.1 == 200 {
             let result_string = String::from_utf8_lossy(&result.0);
             println!("{}", result_string);
-            Some(serde_xml::deserialize(result_string.as_bytes()).context(XmlError)?)
+            Some(serde_xml::deserialize(result_string.as_bytes())?)
         } else {
             None
         };
@@ -398,9 +383,9 @@ impl Bucket {
             continuation_token,
         };
         let request = Request::new(self, "/", command);
-        let result = request.response_data().context(RequestError)?;
+        let result = request.response_data()?;
         let result_string = String::from_utf8_lossy(&result.0);
-        let deserialized: ListBucketResult = serde_xml::deserialize(result_string.as_bytes()).context(XmlError)?;
+        let deserialized: ListBucketResult = serde_xml::deserialize(result_string.as_bytes())?;
         Ok((deserialized, result.1))
     }
 
@@ -501,7 +486,7 @@ impl Bucket {
                 }
             })
         });
-        list_entire_bucket.and_then(|(_token, results): (Option<&str>, Vec<ListBucketResult>)| Ok(results)).map_err(|_| S3Error {})
+        list_entire_bucket.and_then(|(_token, results): (Option<&str>, Vec<ListBucketResult>)| Ok(results)).map_err(|_| S3Error { src: None })
     }
 
 

--- a/src/bucket.rs
+++ b/src/bucket.rs
@@ -388,8 +388,8 @@ impl Bucket {
     }
 
     fn list_page(&self,
-             prefix: &str,
-             delimiter: Option<&str>,
+             prefix: String,
+             delimiter: Option<String>,
              continuation_token: Option<String>)
              -> S3Result<(ListBucketResult, u16)> {
         let command = Command::ListBucket {
@@ -405,8 +405,8 @@ impl Bucket {
     }
 
     pub fn list_page_async(&self,
-                   prefix: &str,
-                   delimiter: Option<&str>,
+                   prefix: String,
+                   delimiter: Option<String>,
                    continuation_token: Option<String>)
                    -> impl Future<Item=(ListBucketResult, u16), Error=S3Error> + Send {
         let command = Command::ListBucket {
@@ -450,13 +450,13 @@ impl Bucket {
     ///     println!("{:?}", list);
     /// }
     /// ```
-    pub fn list_all(&self, prefix: &str, delimiter: Option<&str>) -> S3Result<Vec<(ListBucketResult, u16)>> {
+    pub fn list_all(&self, prefix: String, delimiter: Option<String>) -> S3Result<Vec<(ListBucketResult, u16)>> {
         let mut results = Vec::new();
-        let mut result = self.list_page(prefix, delimiter, None)?;
+        let mut result = self.list_page(prefix.clone(), delimiter.clone(), None)?;
         loop {
             results.push(result.clone());
             match result.0.next_continuation_token {
-                Some(token) => result = self.list_page(prefix, delimiter, Some(token))?,
+                Some(token) => result = self.list_page(prefix.clone(), delimiter.clone(), Some(token))?,
                 None => break,
             }
         }
@@ -489,10 +489,10 @@ impl Bucket {
     ///     Ok(())
     /// });
     /// ```
-    pub fn list_all_async<'a>(&'a self, prefix: &'a str, delimiter: Option<&'a str>) -> impl Future<Item=Vec<ListBucketResult>, Error=S3Error> + 'a + Send {
+    pub fn list_all_async(&self, prefix: String, delimiter: Option<String>) -> impl Future<Item=Vec<ListBucketResult>, Error=S3Error> + Send + '_ {
         let list_entire_bucket = loop_fn((None, Vec::new()), move |(continuation_token, results): (Option<String>, Vec<ListBucketResult>)| {
             let mut inner_results = results;
-            self.list_page_async(prefix, delimiter, continuation_token).and_then(|(result, _status_code)| {
+            self.list_page_async(prefix.clone(), delimiter.clone(), continuation_token).and_then(|(result, _status_code)| {
                 inner_results.push(result.clone());
                 match result.next_continuation_token {
                     Some(token) => Ok(Loop::Continue((Some(token), inner_results))),

--- a/src/bucket.rs
+++ b/src/bucket.rs
@@ -24,7 +24,7 @@ pub enum Error {
     },
     ParseError {
         source: ::region::Error
-    }
+    },
 }
 
 type S3Result<T, E = Error> = std::result::Result<T, E>;
@@ -386,10 +386,10 @@ impl Bucket {
         Ok((tagging, result.1))
     }
 
-    fn _list(&self,
+    fn list_page(&self,
              prefix: &str,
              delimiter: Option<&str>,
-             continuation_token: Option<&str>)
+             continuation_token: Option<String>)
              -> S3Result<(ListBucketResult, u16)> {
         let command = Command::ListBucket {
             prefix,
@@ -403,10 +403,10 @@ impl Bucket {
         Ok((deserialized, result.1))
     }
 
-    fn _list_async(&self,
+    pub fn list_page_async(&self,
                    prefix: &str,
                    delimiter: Option<&str>,
-                   continuation_token: Option<&str>)
+                   continuation_token: Option<String>)
                    -> impl Future<Item=(ListBucketResult, u16)> {
         let command = Command::ListBucket {
             prefix,
@@ -443,19 +443,19 @@ impl Bucket {
     /// let credentials = Credentials::default();
     /// let bucket = Bucket::new(bucket_name, region, credentials).unwrap();
     ///
-    /// let results = bucket.list("/", Some("/")).unwrap();
+    /// let results = bucket.list_all("/", Some("/")).unwrap();
     /// for (list, code) in results {
     ///     assert_eq!(200, code);
     ///     println!("{:?}", list);
     /// }
     /// ```
-    pub fn list(&self, prefix: &str, delimiter: Option<&str>) -> S3Result<Vec<(ListBucketResult, u16)>> {
+    pub fn list_all(&self, prefix: &str, delimiter: Option<&str>) -> S3Result<Vec<(ListBucketResult, u16)>> {
         let mut results = Vec::new();
-        let mut result = self._list(prefix, delimiter, None)?;
+        let mut result = self.list_page(prefix, delimiter, None)?;
         loop {
             results.push(result.clone());
             match result.0.next_continuation_token {
-                Some(token) => result = self._list(prefix, delimiter, Some(&token))?,
+                Some(token) => result = self.list_page(prefix, delimiter, Some(token))?,
                 None => break,
             }
         }
@@ -463,44 +463,45 @@ impl Bucket {
         Ok(results)
     }
 
-//    pub fn list_async(&self, prefix: &str, delimiter: Option<&str>) {
-//        let mut done = false;
-//        let list_entire_bucket_future = loop_fn((None, Vec::new()), |(continuation_token, results): (Option<String>, Vec<ListBucketResult>)|{
-//            let mut inner_results = results;
-//            let mut token: Option<String> = None;
-//            let ct = if continuation_token.is_some() {
-//                Some(continuation_token.unwrap().as_str())
-//            } else {
-//                None
-//            };
-//            let result_future = self._list_async(prefix, delimiter, ct);
-//            result_future.map(|item_future| {
-//                item_future.map(|s3_result| {
-//                    match s3_result {
-//                        Ok((list_bucket_result, status_code)) => {
-//                            inner_results.push(list_bucket_result.clone());
-//                            if let Some(tkn) = list_bucket_result.next_continuation_token {
-//                                token = Some(tkn.to_string());
-//                                done = false
-//                            } else {
-//                               done = true
-//                            }
-//                        },
-//                        Err(e) => panic!(e)
-//                    }
-//                })
-//            });
-//            if done {
-//                Ok(Loop::Break((None, inner_results)))
-//            } else {
-//                Ok(Loop::Continue((token, inner_results)))
-//            }
-//
-//        });
-//
-//
-////        Ok(results);
-//    }
+
+    /// List the contents of an S3 bucket.
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// extern crate futures;
+    /// use std::str;
+    /// use s3::bucket::Bucket;
+    /// use s3::credentials::Credentials;
+    /// use futures::future::Future;
+    /// let bucket_name = &"rust-s3-test";
+    /// let aws_access = &"access_key";
+    /// let aws_secret = &"secret_key";
+    ///
+    /// let bucket_name = &"rust-s3-test";
+    /// let region = "us-east-1".parse().unwrap();
+    /// let credentials = Credentials::default();
+    /// let bucket = Bucket::new(bucket_name, region, credentials).unwrap();
+    ///
+    /// let results = bucket.list_all_async("/", Some("/")).and_then(|list| {
+    ///     println!("{:?}", list);
+    ///     Ok(())
+    /// });
+    /// ```
+    pub fn list_all_async<'a>(&'a self, prefix: &'a str, delimiter: Option<&'a str>) -> impl Future<Item=Vec<ListBucketResult>> + 'a {
+        let list_entire_bucket = loop_fn((None, Vec::new()), move |(continuation_token, results): (Option<String>, Vec<ListBucketResult>)| {
+            let mut inner_results = results;
+            self.list_page_async(prefix, delimiter, continuation_token).and_then(|(result, _status_code)| {
+                inner_results.push(result.clone());
+                match result.next_continuation_token {
+                    Some(token) => Ok(Loop::Continue((Some(token), inner_results))),
+                    None => Ok(Loop::Break((None, inner_results)))
+                }
+            })
+        });
+        list_entire_bucket.and_then(|(_token, results): (Option<&str>, Vec<ListBucketResult>)| Ok(results))
+    }
+
 
     /// Get a reference to the name of the S3 bucket.
     pub fn name(&self) -> String {

--- a/src/bucket.rs
+++ b/src/bucket.rs
@@ -444,7 +444,7 @@ impl Bucket {
     /// let credentials = Credentials::default();
     /// let bucket = Bucket::new(bucket_name, region, credentials).unwrap();
     ///
-    /// let results = bucket.list_all("/", Some("/")).unwrap();
+    /// let results = bucket.list_all("/".to_string(), Some("/".to_string())).unwrap();
     /// for (list, code) in results {
     ///     assert_eq!(200, code);
     ///     println!("{:?}", list);
@@ -484,7 +484,7 @@ impl Bucket {
     /// let credentials = Credentials::default();
     /// let bucket = Bucket::new(bucket_name, region, credentials).unwrap();
     ///
-    /// let results = bucket.list_all_async("/", Some("/")).and_then(|list| {
+    /// let results = bucket.list_all_async("/".to_string(), Some("/".to_string())).and_then(|list| {
     ///     println!("{:?}", list);
     ///     Ok(())
     /// });

--- a/src/bucket.rs
+++ b/src/bucket.rs
@@ -7,7 +7,7 @@ use std::io::Write;
 use credentials::Credentials;
 use command::Command;
 use region::Region;
-use request::{Request, Headers, Query};
+use request::{Request, Headers, Query, S3Error};
 use serde_types::{ListBucketResult, BucketLocationResult};
 use serde_types::Tagging;
 use futures::Future;
@@ -25,6 +25,7 @@ pub enum Error {
     ParseError {
         source: ::region::Error
     },
+    ResponseError
 }
 
 type S3Result<T, E = Error> = std::result::Result<T, E>;
@@ -407,7 +408,7 @@ impl Bucket {
                    prefix: &str,
                    delimiter: Option<&str>,
                    continuation_token: Option<String>)
-                   -> impl Future<Item=(ListBucketResult, u16)> {
+                   -> impl Future<Item=(ListBucketResult, u16), Error=S3Error> + Send {
         let command = Command::ListBucket {
             prefix,
             delimiter,
@@ -488,7 +489,7 @@ impl Bucket {
     ///     Ok(())
     /// });
     /// ```
-    pub fn list_all_async<'a>(&'a self, prefix: &'a str, delimiter: Option<&'a str>) -> impl Future<Item=Vec<ListBucketResult>> + 'a {
+    pub fn list_all_async<'a>(&'a self, prefix: &'a str, delimiter: Option<&'a str>) -> impl Future<Item=Vec<ListBucketResult>, Error=S3Error> + 'a + Send {
         let list_entire_bucket = loop_fn((None, Vec::new()), move |(continuation_token, results): (Option<String>, Vec<ListBucketResult>)| {
             let mut inner_results = results;
             self.list_page_async(prefix, delimiter, continuation_token).and_then(|(result, _status_code)| {
@@ -499,7 +500,7 @@ impl Bucket {
                 }
             })
         });
-        list_entire_bucket.and_then(|(_token, results): (Option<&str>, Vec<ListBucketResult>)| Ok(results))
+        list_entire_bucket.and_then(|(_token, results): (Option<&str>, Vec<ListBucketResult>)| Ok(results)).map_err(|_| S3Error {})
     }
 
 

--- a/src/bucket.rs
+++ b/src/bucket.rs
@@ -489,10 +489,11 @@ impl Bucket {
     ///     Ok(())
     /// });
     /// ```
-    pub fn list_all_async(&self, prefix: String, delimiter: Option<String>) -> impl Future<Item=Vec<ListBucketResult>, Error=S3Error> + Send + '_ {
+    pub fn list_all_async(&self, prefix: String, delimiter: Option<String>) -> impl Future<Item=Vec<ListBucketResult>, Error=S3Error> + Send {
+        let the_bucket = self.to_owned();
         let list_entire_bucket = loop_fn((None, Vec::new()), move |(continuation_token, results): (Option<String>, Vec<ListBucketResult>)| {
             let mut inner_results = results;
-            self.list_page_async(prefix.clone(), delimiter.clone(), continuation_token).and_then(|(result, _status_code)| {
+            the_bucket.list_page_async(prefix.clone(), delimiter.clone(), continuation_token).and_then(|(result, _status_code)| {
                 inner_results.push(result.clone());
                 match result.next_continuation_token {
                     Some(token) => Ok(Loop::Continue((Some(token), inner_results))),

--- a/src/command.rs
+++ b/src/command.rs
@@ -16,7 +16,7 @@ pub enum Command<'a> {
     ListBucket {
         prefix: &'a str,
         delimiter: Option<&'a str>,
-        continuation_token: Option<&'a str>
+        continuation_token: Option<String>
     },
     GetBucketLocation
 }

--- a/src/command.rs
+++ b/src/command.rs
@@ -1,5 +1,6 @@
 use reqwest::Method;
 
+#[derive(Clone)]
 pub enum Command<'a> {
     DeleteObject,
     DeleteObjectTagging,
@@ -14,8 +15,8 @@ pub enum Command<'a> {
     },
 
     ListBucket {
-        prefix: &'a str,
-        delimiter: Option<&'a str>,
+        prefix: String,
+        delimiter: Option<String>,
         continuation_token: Option<String>
     },
     GetBucketLocation

--- a/src/credentials.rs
+++ b/src/credentials.rs
@@ -1,30 +1,7 @@
 use std::env;
 use dirs;
 use ini::Ini;
-use snafu::{ResultExt, Snafu};
-
-#[derive(Debug, Snafu)]
-pub enum Error {
-    VarError {
-        source: std::env::VarError
-    },
-    HomeDir,
-    IniError {
-        source: ini::ini::Error
-    },
-    #[snafu(display("Section [{}] not found in {}", section, profile))]
-    AwsMissingSection {
-        section: String,
-        profile: String
-    },
-    #[snafu(display("Missing {} in {}", part, profile))]
-    AwsConfigError {
-        part: String,
-        profile: String
-    }
-}
-
-type S3Result<T, E = Error> = std::result::Result<T, E>;
+use error::{err, S3Result};
 
 /// AWS access credentials: access key, secret key, and optional token.
 ///
@@ -125,8 +102,8 @@ impl Credentials {
     }
 
     fn from_env() -> S3Result<Credentials> {
-        let access_key = env::var("AWS_ACCESS_KEY_ID").context(VarError)?;
-        let secret_key = env::var("AWS_SECRET_ACCESS_KEY").context(VarError)?;
+        let access_key = env::var("AWS_ACCESS_KEY_ID")?;
+        let secret_key = env::var("AWS_SECRET_ACCESS_KEY")?;
         let token = match env::var("AWS_SESSION_TOKEN") {
             Ok(x) => Some(x),
             Err(_) => None
@@ -137,29 +114,29 @@ impl Credentials {
     fn from_profile(section: Option<String>) -> S3Result<Credentials> {
         let home_dir = match dirs::home_dir() {
             Some(path) => Ok(path),
-            None => Err(Error::HomeDir),
+            None => Err(err("Invalid home dir")),
         };
         let profile = format!("{}/.aws/credentials", home_dir?.display());
-        let conf = Ini::load_from_file(&profile).context(IniError)?;
+        let conf = Ini::load_from_file(&profile)?;
         let section = match section {
             Some(s) => s,
             None => String::from("default")
         };
         let data1 = match conf.section(Some(section.clone())) {
             Some(d) => Ok(d),
-            None => Err(Error::AwsMissingSection {section: section.clone(), profile: profile.clone()})
+            None => Err(err("Missing aws section"))
         };
         let data2 = match conf.section(Some(section.clone())) {
             Some(d) => Ok(d),
-            None => Err(Error::AwsMissingSection {section, profile: profile.clone()})
+            None => Err(err("Missing aws section"))
         };
         let access_key = match data1?.get("aws_access_key_id") {
             Some(x) => Ok(x.to_owned()),
-            None => Err(Error::AwsConfigError {part: "aws_access_key_id".to_string(), profile: profile.clone()})
+            None => Err(err("Missing aws section"))
         };
         let secret_key = match data2?.get("aws_secret_access_key") {
             Some(x) => Ok(x.to_owned()),
-            None => Err(Error::AwsConfigError {part: "aws_secret_access_key".to_string(), profile})
+            None => Err(err("Missing aws section"))
         };
         Ok(Credentials { access_key: access_key?.to_owned(), secret_key: secret_key?.to_owned(), token: None, _private: () })
     }

--- a/src/error.rs
+++ b/src/error.rs
@@ -1,0 +1,49 @@
+use std;
+use serde_xml;
+use reqwest;
+use core::fmt;
+use std::error::Error;
+
+macro_rules! impl_from {
+    ($t: ty) => {
+        impl From<$t> for S3Error {
+            fn from(e: $t) -> S3Error {
+                S3Error { src: Some(String::from(format!("{}",e))) }
+            }
+        }
+    }
+}
+
+impl_from!(serde_xml::Error);
+impl_from!(reqwest::Error);
+impl_from!(reqwest::header::InvalidHeaderName);
+impl_from!(reqwest::header::InvalidHeaderValue);
+impl_from!(std::env::VarError);
+impl_from!(ini::ini::Error);
+
+#[derive(Debug)]
+pub struct S3Error {
+    pub src: Option<String>
+}
+
+pub fn err(e: &str) -> S3Error {
+    S3Error {src: Some(String::from(e))}
+}
+
+impl fmt::Display for S3Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self.src.as_ref() {
+            Some(err) => write!(f, "{}", err),
+            None => write!(f, "An unknown error has occured!")
+        }
+
+    }
+}
+
+impl Error for S3Error {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        None
+    }
+}
+
+pub type S3Result<T, E = S3Error> = std::result::Result<T, E>;

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -12,7 +12,6 @@ extern crate url;
 extern crate ini;
 extern crate dirs;
 extern crate futures;
-extern crate snafu;
 extern crate core;
 
 pub mod bucket;
@@ -23,6 +22,8 @@ pub mod request;
 pub mod serde_types;
 pub mod signing;
 pub mod deserializer;
+#[macro_use]
+pub mod error;
 
 const LONG_DATE: &str = "%Y%m%dT%H%M%SZ";
 const EMPTY_PAYLOAD_SHA: &str = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";

--- a/src/region.rs
+++ b/src/region.rs
@@ -1,19 +1,6 @@
 use std::fmt;
 use std::str::{self, FromStr};
-use snafu::Snafu;
-
-#[derive(Debug, Snafu)]
-pub enum Error {
-    XmlError,
-    RequestError {
-        source: ::request::Error
-    },
-    ParseError {
-        source: std::string::ParseError
-    }
-}
-
-type S3Result<T, E = Error> = std::result::Result<T, E>;
+use error::{S3Result, S3Error};
 
 /// AWS S3 [region identifier](https://docs.aws.amazon.com/general/latest/gr/rande.html#s3_region),
 /// passing in custom values is also possible, in that case it is up to you to pass a valid endpoint,
@@ -115,7 +102,7 @@ impl fmt::Display for Region {
 }
 
 impl FromStr for Region {
-    type Err = Error;
+    type Err = S3Error;
 
     fn from_str(s: &str) -> S3Result<Self> {
         use self::Region::*;
@@ -179,7 +166,7 @@ impl Region {
 
     pub fn scheme(&self) -> String {
         match *self {
-            Region::Custom{ region: _, ref endpoint } => {
+            Region::Custom{ ref endpoint, ..} => {
                 match endpoint.find("://") {
                     Some(pos) => endpoint[..pos].to_string(),
                     None => "https".to_string()
@@ -191,7 +178,7 @@ impl Region {
 
     pub fn host(&self) -> String {
         match *self {
-            Region::Custom{ region: _, ref endpoint } => {
+            Region::Custom{ ref endpoint, ..} => {
                 match endpoint.find("://") {
                     Some(pos) => endpoint[pos + 3..].to_string(),
                     None => endpoint.to_string()

--- a/src/request.rs
+++ b/src/request.rs
@@ -24,6 +24,27 @@ use signing;
 use EMPTY_PAYLOAD_SHA;
 use LONG_DATE;
 use reqwest::async::Response;
+use core::fmt;
+use std::error;
+
+#[derive(Debug, Default)]
+pub struct S3Error {}
+
+impl fmt::Display for S3Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "S3Error!")
+    }
+}
+
+impl error::Error for S3Error {
+    fn description(&self) -> &str {
+        "Description for ErrorB"
+    }
+
+    fn cause(&self) -> Option<&error::Error> {
+        None
+    }
+}
 
 #[derive(Debug, Snafu)]
 pub enum Error {
@@ -34,6 +55,7 @@ pub enum Error {
         source: reqwest::header::InvalidHeaderValue
     },
     ReqwestFuture,
+    ResponseError,
     ParseError {
         source: std::string::ParseError
     }
@@ -265,7 +287,7 @@ impl<'a> Request<'a> {
         }
     }
 
-    pub fn response_future(&self) -> impl Future<Item=Response, Error=reqwest::Error> {
+    pub fn response_future(&self) -> impl Future<Item=Response, Error=S3Error> {
         let client = if cfg!(feature = "no-verify-ssl") {
             async::Client::builder()
                 .danger_accept_invalid_certs(true)
@@ -292,7 +314,7 @@ impl<'a> Request<'a> {
             .headers(headers.to_owned())
             .body(content.to_owned());
 
-        request.send()
+        request.send().map_err(|_| S3Error {})
     }
 
 //    pub fn response_data_future2(&self) -> impl Future<Item=S3Result<(impl Future<Item=Vec<u8>>, u16)>> {
@@ -304,11 +326,11 @@ impl<'a> Request<'a> {
 //        })
 //    }
 
-    pub fn response_data_future(&self) -> impl Future<Item=(Vec<u8>, u16)> {
+    pub fn response_data_future(&self) -> impl Future<Item=(Vec<u8>, u16), Error=S3Error> {
         self.response_future()
-            .and_then(|mut response| Ok((response.text(), response.status().as_u16())))
+            .and_then(|mut response| Ok((response.text(), response.status().as_u16()))).map_err(|_| S3Error {})
             .and_then(|(body_future, status_code)| {
-                body_future.and_then(move |body| Ok((body.as_bytes().to_vec(), status_code)))
+                body_future.and_then(move |body| Ok((body.as_bytes().to_vec(), status_code))).map_err(|_| S3Error {})
             })
     }
 

--- a/src/request.rs
+++ b/src/request.rs
@@ -1,8 +1,6 @@
 extern crate base64;
 extern crate md5;
 
-use snafu::{ResultExt, Snafu};
-
 use std::collections::HashMap;
 use std::io::{Read, Write};
 
@@ -18,61 +16,13 @@ use url::Url;
 
 use futures::prelude::*;
 
-use serde_types::AwsError;
+//use serde_types::AwsError;
 use signing;
 
 use EMPTY_PAYLOAD_SHA;
 use LONG_DATE;
 use reqwest::async::Response;
-use core::fmt;
-use std::error;
-
-#[derive(Debug, Default)]
-pub struct S3Error {}
-
-impl fmt::Display for S3Error {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "S3Error!")
-    }
-}
-
-impl error::Error for S3Error {
-    fn description(&self) -> &str {
-        "Description for S3Error"
-    }
-
-    fn cause(&self) -> Option<&error::Error> {
-        None
-    }
-}
-
-#[derive(Debug, Snafu)]
-pub enum Error {
-    InvalidHeaderName {
-        source: reqwest::header::InvalidHeaderName
-    },
-    InvalidHeaderValue {
-        source: reqwest::header::InvalidHeaderValue
-    },
-    ReqwestFuture,
-    ResponseError,
-    ParseError {
-        source: std::string::ParseError
-    }
-
-}
-
-type S3Result<T, E = Error> = std::result::Result<T, E>;
-
-#[derive(Debug, Snafu)]
-enum InternalError {
-    BucketError {
-        source: ::bucket::Error
-    },
-    RegionError {
-        source: ::region::Error
-    }
-}
+use error::{S3Error, S3Result, err};
 
 /// Collection of HTTP headers sent to S3 service, in key/value format.
 pub type Headers = HashMap<String, String>;
@@ -234,33 +184,33 @@ impl<'a> Request<'a> {
             .bucket
             .extra_headers
             .iter()
-            .map(|(k, v)| Ok((k.parse::<HeaderName>().context(InvalidHeaderName)?, v.parse::<HeaderValue>().context(InvalidHeaderValue)?)))
-            .collect::<Result<HeaderMap, Error>>()?;
+            .map(|(k, v)| Ok((k.parse::<HeaderName>()?, v.parse::<HeaderValue>()?)))
+            .collect::<Result<HeaderMap, S3Error>>()?;
         match self.command {
-            Command::GetBucketLocation => headers.insert(header::HOST, self.bucket.self_host().parse().context(InvalidHeaderValue)?),
-            _ => headers.insert(header::HOST, self.bucket.host().parse().context(InvalidHeaderValue)?)
+            Command::GetBucketLocation => headers.insert(header::HOST, self.bucket.self_host().parse()?),
+            _ => headers.insert(header::HOST, self.bucket.host().parse()?)
         };
         headers.insert(
             header::CONTENT_LENGTH,
-            self.content_length().to_string().parse().context(InvalidHeaderValue)?,
+            self.content_length().to_string().parse()?,
         );
-        headers.insert(header::CONTENT_TYPE, self.content_type().parse().context(InvalidHeaderValue)?);
-        headers.insert("X-Amz-Content-Sha256", sha256.parse().context(InvalidHeaderValue)?);
-        headers.insert("X-Amz-Date", self.long_date().parse().context(InvalidHeaderValue)?);
+        headers.insert(header::CONTENT_TYPE, self.content_type().parse()?);
+        headers.insert("X-Amz-Content-Sha256", sha256.parse()?);
+        headers.insert("X-Amz-Date", self.long_date().parse()?);
 
         if let Some(token) = self.bucket.credentials().token.as_ref() {
-            headers.insert("X-Amz-Security-Token", token.parse().context(InvalidHeaderValue)?);
+            headers.insert("X-Amz-Security-Token", token.parse()?);
         }
 
         if let Command::PutObjectTagging { tags } = self.command {
             let digest = md5::compute(tags);
             let hash = base64::encode(digest.as_ref());
-            headers.insert("Content-MD5", hash.parse().context(InvalidHeaderValue)?);
+            headers.insert("Content-MD5", hash.parse()?);
         }
 
         // This must be last, as it signs the other headers
         let authorization = self.authorization(&headers);
-        headers.insert(header::AUTHORIZATION, authorization.parse().context(InvalidHeaderValue)?);
+        headers.insert(header::AUTHORIZATION, authorization.parse()?);
 
         // The format of RFC2822 is somewhat malleable, so including it in
         // signed headers can cause signature mismatches. We do include the
@@ -268,7 +218,7 @@ impl<'a> Request<'a> {
         // range and can't be used again e.g. reply attacks. Adding this header
         // after the generation of the Authorization header leaves it out of
         // the signed headers.
-        headers.insert(header::DATE, self.datetime.to_rfc2822().parse().context(InvalidHeaderValue)?);
+        headers.insert(header::DATE, self.datetime.to_rfc2822().parse()?);
 
         Ok(headers)
     }
@@ -276,14 +226,14 @@ impl<'a> Request<'a> {
     pub fn response_data(&self) -> S3Result<(Vec<u8>, u16)> {
         match self.response_data_future().wait() {
             Ok((response_data, status_code)) => Ok((response_data, status_code)),
-            Err(_) => Err(Error::ReqwestFuture)
+            Err(_) => Err(S3Error { src: Some("Error getting response".to_string())})
         }
     }
 
     pub fn response_data_to_writer<T: Write>(&self, writer: &mut T) -> S3Result<u16> {
         match self.response_data_to_writer_future(writer).wait() {
             Ok(status_code) => Ok(status_code),
-            Err(_) => Err(Error::ReqwestFuture)
+            Err(_) => Err(err("ReqwestFuture"))
         }
     }
 
@@ -314,14 +264,14 @@ impl<'a> Request<'a> {
             .headers(headers.to_owned())
             .body(content.to_owned());
 
-        request.send().map_err(|_| S3Error {})
+        request.send().map_err(|_| S3Error {src: None})
     }
 
     pub fn response_data_future(&self) -> impl Future<Item=(Vec<u8>, u16), Error=S3Error> {
         self.response_future()
-            .and_then(|mut response| Ok((response.text(), response.status().as_u16()))).map_err(|_| S3Error {})
+            .and_then(|mut response| Ok((response.text(), response.status().as_u16()))).map_err(|_| S3Error {src: None})
             .and_then(|(body_future, status_code)| {
-                body_future.and_then(move |body| Ok((body.as_bytes().to_vec(), status_code))).map_err(|_| S3Error {})
+                body_future.and_then(move |body| Ok((body.as_bytes().to_vec(), status_code))).map_err(|_| S3Error {src: None})
             })
     }
 
@@ -340,11 +290,9 @@ mod tests {
     use bucket::Bucket;
     use command::Command;
     use credentials::Credentials;
-    use request::{Request, InternalError};
-    use request::S3Result;
-    use snafu::ResultExt;
+    use request::{Request};
     use url::form_urlencoded::Parse;
-    use request::{ParseError, BucketError, RegionError};
+    use error::S3Result;
 
     // Fake keys - otherwise using Credentials::default will use actual user
     // credentials if they exist.
@@ -355,9 +303,9 @@ mod tests {
     }
 
     #[test]
-    fn url_uses_https_by_default() -> Result<(), InternalError> {
-        let region = "custom-region".parse().context(RegionError)?;
-        let bucket = Bucket::new("my-first-bucket", region, fake_credentials()).context(BucketError)?;
+    fn url_uses_https_by_default() -> S3Result<()> {
+        let region = "custom-region".parse()?;
+        let bucket = Bucket::new("my-first-bucket", region, fake_credentials())?;
         let path = "/my-first/path";
         let request = Request::new(&bucket, path, Command::GetObject);
 
@@ -371,9 +319,9 @@ mod tests {
     }
 
     #[test]
-    fn url_uses_scheme_from_custom_region_if_defined() -> Result<(), InternalError> {
-        let region = "http://custom-region".parse().context(RegionError)?;
-        let bucket = Bucket::new("my-second-bucket", region, fake_credentials()).context(BucketError)?;
+    fn url_uses_scheme_from_custom_region_if_defined() -> S3Result<()> {
+        let region = "http://custom-region".parse()?;
+        let bucket = Bucket::new("my-second-bucket", region, fake_credentials())?;
         let path = "/my-second/path";
         let request = Request::new(&bucket, path, Command::GetObject);
 

--- a/src/request.rs
+++ b/src/request.rs
@@ -108,13 +108,13 @@ impl<'a> Request<'a> {
             url.query_pairs_mut().append_pair(key, value);
         }
 
-        if let Command::ListBucket { prefix, delimiter, continuation_token } = self.command {
+        if let Command::ListBucket { prefix, delimiter, continuation_token } = &self.command {
             let mut query_pairs = url.query_pairs_mut();
             delimiter.map(|d| query_pairs.append_pair("delimiter", d));
             query_pairs.append_pair("prefix", prefix);
             query_pairs.append_pair("list-type", "2");
             if let Some(token) = continuation_token {
-                query_pairs.append_pair("continuation-token", token);
+                query_pairs.append_pair("continuation-token", &token);
             }
         }
 

--- a/src/request.rs
+++ b/src/request.rs
@@ -38,7 +38,7 @@ impl fmt::Display for S3Error {
 
 impl error::Error for S3Error {
     fn description(&self) -> &str {
-        "Description for ErrorB"
+        "Description for S3Error"
     }
 
     fn cause(&self) -> Option<&error::Error> {
@@ -130,10 +130,10 @@ impl<'a> Request<'a> {
             url.query_pairs_mut().append_pair(key, value);
         }
 
-        if let Command::ListBucket { prefix, delimiter, continuation_token } = &self.command {
+        if let Command::ListBucket { prefix, delimiter, continuation_token } = self.command.clone() {
             let mut query_pairs = url.query_pairs_mut();
-            delimiter.map(|d| query_pairs.append_pair("delimiter", d));
-            query_pairs.append_pair("prefix", prefix);
+            delimiter.map(|d| query_pairs.append_pair("delimiter", &d.clone()));
+            query_pairs.append_pair("prefix", &prefix);
             query_pairs.append_pair("list-type", "2");
             if let Some(token) = continuation_token {
                 query_pairs.append_pair("continuation-token", &token);
@@ -317,15 +317,6 @@ impl<'a> Request<'a> {
         request.send().map_err(|_| S3Error {})
     }
 
-//    pub fn response_data_future2(&self) -> impl Future<Item=S3Result<(impl Future<Item=Vec<u8>>, u16)>> {
-//        let response_future = self.response_future();
-//        response_future.map(|mut response| {
-//            let response_code = response.status().as_u16();
-//            let response_data = response.text().map(|body| body.as_bytes().to_vec());
-//            Ok((response_data, response_code))
-//        })
-//    }
-
     pub fn response_data_future(&self) -> impl Future<Item=(Vec<u8>, u16), Error=S3Error> {
         self.response_future()
             .and_then(|mut response| Ok((response.text(), response.status().as_u16()))).map_err(|_| S3Error {})
@@ -342,18 +333,6 @@ impl<'a> Request<'a> {
         })
     }
 
-//    pub fn response_data_to_writer_future2<'b, T: Write>(&self, writer: &'b mut T) -> impl Future<Item=S3Result<(impl Future<Item=S3Result<()>> + 'b, u16)>> + 'b {
-//        let future_response = self.response_data_future();
-//        future_response.map(|response| {
-//            match response {
-//                Ok(success) => {
-//                    let to_writer_future = success.0.map(move |body| Ok(writer.write_all(body.as_slice())?));
-//                    Ok((to_writer_future, success.1))
-//                }
-//                Err(e) => Err(e)
-//            }
-//        })
-//    }
 }
 
 #[cfg(test)]

--- a/src/signing.rs
+++ b/src/signing.rs
@@ -20,7 +20,7 @@ pub fn uri_encode(string: &str, encode_slash: bool) -> String {
     let mut result = String::with_capacity(string.len() * 2);
     for c in string.chars() {
         match c {
-            'a'...'z' | 'A'...'Z' | '0'...'9' | '_' | '-' | '~' | '.' => result.push(c),
+            'a'..='z' | 'A'..='Z' | '0'..='9' | '_' | '-' | '~' | '.' => result.push(c),
             '/' if encode_slash => result.push_str("%2F"),
             '/' if !encode_slash => result.push('/'),
             _ => {

